src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,1045 @@
+template const Type* TypeIntHelper::int_type_narrow(const TypeLong* new_type, const TypeLong* old_type);
+
+
+#ifndef PRODUCT
+template <class T>
+static const char* int_name_near(T origin, const char* xname, char* buf, size_t buf_size, T n) {
+  if (n < origin) {
+    if (n <= origin - 10000) {
+      return nullptr;
+    }
+    os::snprintf_checked(buf, buf_size, "%s-" INT32_FORMAT, xname, jint(origin - n));
+  } else if (n > origin) {
+    if (n >= origin + 10000) {
+      return nullptr;
+    }
+    os::snprintf_checked(buf, buf_size, "%s+" INT32_FORMAT, xname, jint(n - origin));
+  } else {
+    return xname;
+  }
+  return buf;
+}
+
+const char* TypeIntHelper::intname(char* buf, size_t buf_size, jint n) {
+  const char* str = int_name_near<jint>(max_jint, "maxint", buf, buf_size, n);
+  if (str != nullptr) {
+    return str;
+  }
+
+  str = int_name_near<jint>(min_jint, "minint", buf, buf_size, n);
+  if (str != nullptr) {
+    return str;
+  }
+
+  os::snprintf_checked(buf, buf_size, INT32_FORMAT, n);
+  return buf;
+}
+
+const char* TypeIntHelper::uintname(char* buf, size_t buf_size, juint n) {
+  const char* str = int_name_near<juint>(max_juint, "maxuint", buf, buf_size, n);
+  if (str != nullptr) {
+    return str;
+  }
+
+  str = int_name_near<juint>(max_jint, "maxint", buf, buf_size, n);
+  if (str != nullptr) {
+    return str;
+  }
+
+  os::snprintf_checked(buf, buf_size, UINT32_FORMAT"u", n);
+  return buf;
+}
+
+const char* TypeIntHelper::longname(char* buf, size_t buf_size, jlong n) {
+  const char* str = int_name_near<jlong>(max_jlong, "maxlong", buf, buf_size, n);
+  if (str != nullptr) {
+    return str;
+  }
+
+  str = int_name_near<jlong>(min_jlong, "minlong", buf, buf_size, n);
+  if (str != nullptr) {
+    return str;
+  }
+
+  str = int_name_near<jlong>(max_juint, "maxuint", buf, buf_size, n);
+  if (str != nullptr) {
+    return str;
+  }
+
+  str = int_name_near<jlong>(max_jint, "maxint", buf, buf_size, n);
+  if (str != nullptr) {
+    return str;
+  }
+
+  str = int_name_near<jlong>(min_jint, "minint", buf, buf_size, n);
+  if (str != nullptr) {
+    return str;
+  }
+
+  os::snprintf_checked(buf, buf_size, JLONG_FORMAT, n);
+  return buf;
+}
+
+const char* TypeIntHelper::ulongname(char* buf, size_t buf_size, julong n) {
+  const char* str = int_name_near<julong>(max_julong, "maxulong", buf, buf_size, n);
+  if (str != nullptr) {
+    return str;
+  }
+
+  str = int_name_near<julong>(max_jlong, "maxlong", buf, buf_size, n);
+  if (str != nullptr) {
+    return str;
+  }
+
+  str = int_name_near<julong>(max_juint, "maxuint", buf, buf_size, n);
+  if (str != nullptr) {
+    return str;
+  }
+
+  str = int_name_near<julong>(max_jint, "maxint", buf, buf_size, n);
+  if (str != nullptr) {
+    return str;
+  }
+
+  os::snprintf_checked(buf, buf_size, JULONG_FORMAT"u", n);
+  return buf;
+}
+
+template <class U>
+const char* TypeIntHelper::bitname(char* buf, size_t buf_size, U zeros, U ones) {
```

Comment:
Wow, these will look incredibly long for 64bit long values, no?
Well, if it ever gets too much in the way, we can still try to find more compressed representations later.
Maybe things like: `*..*000` for 8-aligned values.
Others: `0..0***`, `00*..*`, `1..1***` etc.
`0..0*..*` would be a little unfortunate as we would lose the position where the bits flip.

Reply:
Hmm, maybe we can find some way to only print the bits if they give "additional information" that is not given by the ranges. Otherwise, we would not print alignment information, and that's quite a shame.

Reply:
Maybe that's worth thinking about, the thing is that bit information is exceptionally long and may not be universally interesting. I think we can think more about it later. My rough idea is that it may be better to print it when it is interesting. For example, we can print alignment properties at the memory access.

Reply:
Yes, we can always improve printing in a later RFE :)